File: app/static/jsx/problem-generator.js
import _ from 'lodash';


function genProblem(ptype, difficulty) {
    const problems = {
        'Addition': addProblem,
        'Subtraction': subProblem,
        'Multiplication': multProblem,
        'Division': divProblem,
        'Hexadecimal': hexProblem,
    }

    return problems[ptype](difficulty);
}

function getDigits(difficulty) {
    const difA = Math.floor(difficulty / 2) + 1;
    const difB = Math.floor((difficulty - 1) / 2) + 1;

    return [difA, difB];
}

function addProblem(difficulty) {
    const firstOperand = Math.round(Math.random() * 100 * 10 ** difficulty);
    const secondOperand = Math.round(Math.random() * 100 * 10 ** difficulty);

    const question = `${firstOperand} + ${secondOperand}`;
    const answer = firstOperand + secondOperand;

    return [question, answer];
}

function subProblem(difficulty) {
    const firstOperand = Math.round(Math.random() * 100 * 10 ** difficulty);
    const secondOperand = Math.round(Math.random() * 100 * 10 ** difficulty);

    const question = `${firstOperand} - ${secondOperand}`;
    const answer = firstOperand - secondOperand;

    return [question, answer];
}

function multProblem(difficulty) {
    const diffs = getDigits(difficulty);
    const firstPlace = Math.round(Math.random());
    const difA = diffs[firstPlace];
    const difB = diffs[1 - firstPlace];
    const firstOperand = Math.round(Math.random() * 10 ** difA);
    const secondOperand = Math.round(Math.random() * 10 ** difB);

    const question = `${firstOperand} * ${secondOperand}`;
    const answer = firstOperand * secondOperand;

    return [question, answer];
}

function divProblem(difficulty) {
    const diffs = getDigits(difficulty);
    const firstPlace = Math.round(Math.random());
    const difA = diffs[firstPlace];
    const difB = diffs[1 - firstPlace];
    const firstOperand = Math.round(Math.random() * 10 ** difA);
    const secondOperand = Math.round(Math.random() * 10 ** difB);

    const question = `${firstOperand} / ${secondOperand}`;
    const answer = Math.round(firstOperand / secondOperand * 1000) / 1000;

    return [question, answer];
}

function hexProblem(difficulty) {
    let num;
    let targetType;
    let answer;
    const hexMap = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'];
    const direction = Math.random();
    
    if (direction > 0.5) {
        num = _.reduce(_.range(difficulty), (result, n) => {
            const randomHex = hexMap[Math.floor(Math.random() * 16)];
            return result + randomHex;
        }, '');

        targetType = 'decimal';
        let power = num.length - 1;
        answer = 0;

        for(let i = 0; i < num.length; i++) {
            const increase = hexMap.indexOf(num[i]) * 16 ** power;
            answer = answer + increase;
            power = power - 1;
        }
        answer = String(answer);
    } else {
        let remainder;
        num = Math.floor(Math.random() * (10 ** (difficulty + 1)));
        let tempNum = num;
        targetType = 'hexadecimal';
        answer = '';
        while (tempNum / 16 >= 1) {
            remainder = tempNum % 16;
            answer = hexMap[remainder] + answer;
            tempNum = Math.floor(tempNum / 16);
        }
        answer = hexMap[tempNum] + answer;
    }

    const question = `convert ${num} to ${targetType}`;

    return [question, answer];
}



export {
    genProblem,
}

File: app/math_teasers/problem_generator.py
import random
from collections import defaultdict
from app.math_teasers.util import generate_numbers


def addprob(difficulty):
    difficulty = int(difficulty)
    a = random.randint(-100*10**difficulty, 100*10**difficulty)
    b = random.randint(-100*10**difficulty, 100*10**difficulty)
    question = "{} + {}\n".format(a, b)
    answer = float(a + b)
    return question, answer

def generate_hex_problem(difficulty):
    hexmap = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F']
    direction = random.random()
    if direction > 0.5:
        num = "".join([hexmap[random.randint(0, 15)] for _ in range(difficulty)])
        target_type = "decimal"
        power = len(num) - 1
        correct = 0
        for c in num:
            increase = hexmap.index(c) * 16 ** power
            correct = correct + increase
            power = power - 1
    else:
        num = random.randint(0, 10 ** (difficulty + 1))
        tempnum = num
        target_type = "hexadecimal"
        correct = ''
        while tempnum / 16 >= 1:
            rem = int(tempnum % 16)
            correct = hexmap[rem] + correct
            tempnum = tempnum // 16
        correct = hexmap[tempnum] + correct

    return num, target_type, correct


def generate_problem(problem, difficulty):
    problem_map = {
        'add': Add_Problem(difficulty),
        'sub': Sub_Problem(difficulty),
        'mult': Mult_Problem(difficulty),
        'div': Div_Problem(difficulty),
    }
    problem_map_defaulted = defaultdict(lambda: None, problem_map)

    return problem_map_defaulted[problem]


class Problem:
    def __init__(self, difficulty):
        self.difficulty = int(difficulty)


class Add_Problem(Problem):
    def __init__(self, difficulty):
        Problem.__init__(self, difficulty)
        self.first_operand, self.second_operand = generate_numbers('add', self.difficulty)
        self.question = "{} + {}\n".format(self.first_operand, self.second_operand)
        self.operator = '+'
        self.answer = float(self.first_operand + self.second_operand)

class Sub_Problem(Problem):
    def __init__(self, difficulty):
        Problem.__init__(self, difficulty)
        self.first_operand, self.second_operand = generate_numbers('sub', self.difficulty)
        self.question = "{} - {}\n".format(self.first_operand, self.second_operand)
        self.operator = '-'
        self.answer = float(self.first_operand - self.second_operand)

class Mult_Problem(Problem):
    def __init__(self, difficulty):
        Problem.__init__(self, difficulty)
        self.first_operand, self.second_operand = generate_numbers('mult', self.difficulty)
        self.question = "{} * {}\n".format(self.first_operand, self.second_operand)
        self.operator = '*'
        self.answer = float(self.first_operand * self.second_operand)

class Div_Problem(Problem):
    def __init__(self, difficulty):
        Problem.__init__(self, difficulty)
        self.first_operand, self.second_operand = generate_numbers('div', self.difficulty)
        self.question = "{} / {}\n".format(self.first_operand, self.second_operand)
        self.operator = '/'
        self.answer = round(self.first_operand / self.second_operand, 4)

class Hex_Problem(Problem):
    def __init__(self, difficulty):
        Problem.__init__(self, difficulty)
        self.first_operand, self.target_type, self.answer = generate_hex_problem(self.difficulty)
        self.question = "Convert {} to {}\n".format(self.first_operand, self.target_type)

        # attempt = int(inputu) if direction > 0.5 else str(inputu).upper()





